src/interaction/DihedralHarmonic.hpp

```diff
@@ -0,0 +1,201 @@
+
+        real rijjk_sqr = rijjk.sqr();
+        real rjkkn_sqr = rjkkn.sqr();
+
+        real rijjk_abs = sqrt(rijjk_sqr);
+        real rjkkn_abs = sqrt(rjkkn_sqr);
+
+        real inv_rijjk = 1.0 / rijjk_abs;
+        real inv_rjkkn = 1.0 / rjkkn_abs;
+
+        // cosine between planes
+        real cos_phi = (rijjk * rjkkn) * (inv_rijjk * inv_rjkkn);
+        real _phi = acos(cos_phi);
+        if (cos_phi > 1.0) {
+          cos_phi = 1.0;
+          _phi = 1e-10; //not 0.0, because 1.0/sin(_phi) would cause a singularity
+        } else if (cos_phi < -1.0) {
+          cos_phi = -1.0;
+          _phi = M_PI-1e-10;
+        }
+
+        /// The part of the formula. 1/sin(phi) * d/dphi U(phi)
+        /// where the \f$U(\phi) = 0.5 K (\phi_{ijkn} - \phi_0)^2]\f$
+        /// The n is multiplicity.
+        ///
+        /// The derivative of \f$U(phi)\f$ is \f$K*n*sin(\phi_0 - n*\phi)\f$
+        ///
+        real coef1 = (1.0/sin(_phi)) * (K * (_phi - phi0) + 0.5 * K * phi0 * phi0);
+
+        real A1 = inv_rijjk * inv_rjkkn;
+        real A2 = inv_rijjk * inv_rijjk;
+        real A3 = inv_rjkkn * inv_rjkkn;
+
+        Real3D p3232 ( prod(r32,r32) );
+        Real3D p3243 ( prod(r32,r43) );
+        Real3D p2132 ( prod(r21,r32) );
+        Real3D p2143 ( prod(r21,r43) );
+        Real3D p2121 ( prod(r21,r21) );
+        Real3D p4343 ( prod(r43,r43) );
+
+        // we have 4 particles 1,2,3,4
+        for(int l=0; l<4; l++){
+          Real3D B1, B2, B3;
+
+          for(int i=0; i<3; i++){
+            B1[i]= r21[i] * ( p3232[i] * (d(l,2)-d(l,3)) + p3243[i] * (d(l,2)-d(l,1)) ) +
+                   r32[i] * ( p2132[i] * (d(l,3)-d(l,2)) + p3243[i] * (d(l,1)-d(l,0)) ) +
+                   r43[i] * ( p2132[i] * (d(l,2)-d(l,1)) + p3232[i] * (d(l,0)-d(l,1)) ) +
+                   2.0 * r32[i] * p2143[i] * (d(l,1)-d(l,2));
+
+            B2[i]= 2.0 * r21[i] * ( p3232[i] * (d(l,1)-d(l,0)) + p2132[i] * (d(l,1)-d(l,2)) ) +
+                   2.0 * r32[i] * ( p2121[i] * (d(l,2)-d(l,1)) + p2132[i] * (d(l,0)-d(l,1)));
+
+            B3[i]= 2.0 * r43[i] * ( p3232[i] * (d(l,3)-d(l,2)) + p3243[i] * (d(l,1)-d(l,2)) ) +
+                   2.0 * r32[i] * ( p4343[i] * (d(l,2)-d(l,1)) + p3243[i] * (d(l,2)-d(l,3)));
+          }
+
+          retF[l] = coef1 * ( A1*B1 - 0.5*cos_phi*(A2*B2 + A3*B3) );
+        }
+
+        force1 = retF[0];
+        force2 = retF[1];
+        force3 = retF[2];
+        force4 = retF[3];
+      }
+
+      /*
+       * Compute raw force for the tabulated potential, depend only on the phi value
+       * \f$F(\phi) = -grad V = -grad [ 0.5 (\phi_{ijkn} - \phi_0)^2] ]  \f$
+       * where n is a multiplicity and \f$phi_0\f$ is the reference angle
+       *
+       * @param[in] phi Degree in radians
+       *
+       * @return The value of the force
+       */
+      real _computeForceRaw(real phi) const {
+	real coef1 = K * (phi - phi0) + 0.5 * K * phi0 * phi0;
+        return -1.0 * coef1 ;
```

Comment:
I've looked at TabulatedDihedral and I don't see any prefactor so we should include it in _computeForceRaw to be internally consistent. 
I am not sure about minus sign.. code is a bit obfuscated.